tests/cpp_unit_tests/test_sparse_idx_vector.cpp

```diff
@@ -0,0 +1,92 @@
+    // Sparse vector to test
+    IdxVector const sample_indptr{0, 0, 3, 3, 6, 7, 7};
+    IdxVector expected_elements{0, 1, 2, 3, 4, 5, 6};
+    std::vector<IdxCount> expected_idx_counts_groups{expected_elements.begin(), expected_elements.end()};
+    SparseIdxVector sparse_idx_vector{sample_indptr};
+
+    // 2nd sparse vector
+    IdxVector const sample_indptr_2{0, 0, 1, 3, 6, 6, 6};
+    IdxVector const expected_elements{0, 1, 2, 3, 4, 5};
+    std::vector<IdxCount> expected_idx_counts_groups_2{expected_elements.begin(), expected_elements.end()};
+    SparseIdxVector sparse_idx_vector_2{sample_indptr_2};
+
+    // Dense Vector
+    IdxVector sample_dense_vector{1, 1, 1, 3, 3, 3, 4};
+    DenseIdxVector dense_idx_vector{sample_dense_vector};
+
+    SUBCASE("Sparse Idx vector") {
+        // Check each group individually
+        std::vector<IdxCount> actual_idx_counts{};
+        for (size_t group_number = 0; group_number < 6; group_number++) {
+            // Prepare values for single group
+            size_t range_size = sample_indptr[group_number + 1] - sample_indptr[group_number];
+            actual_idx_counts.clear();
+            actual_idx_counts.resize(range_size);
+            auto group_i = sparse_idx_vector.get_element_range(group_number);
+            std::copy(group_i.begin(), group_i.end(), actual_idx_counts.begin());
+            // Test values for a single group
+            if (range_size == 0) {
+                CHECK(actual_idx_counts.empty());
+            } else {
+                CHECK(actual_idx_counts.front() == IdxCount{sample_indptr[group_number]});
+                CHECK(actual_idx_counts.back() == IdxCount{sample_indptr[group_number + 1] - 1});
+            }
+        }
+
+        // Check iteration for all groups
+        std::vector<IdxCount> actual_idx_counts_groups{};
+        for (auto element_range : sparse_idx_vector) {
+            for (auto& element : element_range) {
+                actual_idx_counts_groups.push_back(element);
+            }
+        }
+        CHECK(actual_idx_counts_groups == expected_idx_counts_groups);
+    }
+
+    SUBCASE("Dense Idx vector") {
+        // Check each group individually
+        std::vector<IdxCount> actual_idx_counts{};
+        for (size_t group_number = 0; group_number < 6; group_number++) {
+            size_t range_size = sample_indptr[group_number + 1] - sample_indptr[group_number];
+            actual_idx_counts.clear();
+            actual_idx_counts.resize(range_size);
+            auto group_i = dense_idx_vector.get_element_range(group_number);
+            std::copy(group_i.begin(), group_i.end(), actual_idx_counts.begin());
+            if (range_size == 0) {
+                CHECK(actual_idx_counts.empty());
+            } else {
+                CHECK(actual_idx_counts.front() == IdxCount{sample_indptr[group_number]});
+                CHECK(actual_idx_counts.back() == IdxCount{sample_indptr[group_number + 1] - 1});
+            }
+        }
+    }
```

Comment:
the test cases for the `SparseIdxVector` and `DenseIdxVector` should IMO be the same aside from the input values. Maybe you can reflect that by making a template test function for that?